ckanext/iotrans/iotrans.py

```diff
@@ -260,16 +260,14 @@ def to_file(context, data_dict):
                             driver=drivers[target_format],
                             crs=from_epsg(target_epsg),
                         ) as outlayer:
-                            outlayer.writerecords(
-                                utils.dump_to_geospatial_generator(
-                                    dump_filepath,
-                                    fieldnames,
-                                    target_format,
-                                    data_dict["source_epsg"],
-                                    target_epsg,
-                                )
-                            )
-                            outlayer.close()
+                            for feature in utils.dump_to_geospatial_generator(
+                                dump_filepath,
+                                fieldnames,
+                                target_format,
+                                data_dict["source_epsg"],
+                                target_epsg,
+                            ):
+                                outlayer.write(feature)
```

Comment:
Just wondering about the switch from `.writerecords` to `.write` here.

I think perhaps motivated by the section of the fiona docs that mention transactions here: https://fiona.readthedocs.io/en/stable/manual.html `ctrl + F` - ` Transactions` (sorry can't link!)

But this paragraph from the docs seems to suggest `.write` may result in more overhead for transactions (1 call to `.write` = 1 new transaction).

Is there something I'm missing here?

> Depending on the driver, a transaction can be a very costly operation. Since write() is just a thin convenience wrapper that calls writerecords() with a single record, you may experience significant performance issue if you write lots of features one by one using this method. Consider preparing your data first and then writing it in a single call to writerecords().

Reply:
@JamieBeverley The idea of using write came from observing complaints from other users about Fiona's writerecords consuming significant memory, growing from 2 GiB to even more. One of the related issues is documented here: [Toblerity/Fiona#685](https://github.com/Toblerity/Fiona/issues/685).

I’m also fully aware that .write(record) is essentially equivalent to .writerecords([record]). This approach writes one record at a time, sacrificing speed to optimize memory usage (which, as I recall, was mentioned in the issue ticket created for this project). Fiona's documentation also states that there’s a trade-off between memory usage and speed.

Regarding the reference note in your comment, I understand that "performance" can refer to various aspects, but I believe the note specifically addresses speed.

That said, we still need to properly monitor resource usage and compare the speed and memory consumption of saving individual records versus batch records. To facilitate this, it would be helpful if you could provide a large CSV file from your end. In the meantime, we can use psutil to measure memory usage when saving single records versus batch records.

Reply:
Thanks @steveoni!

Regarding that linked issue (https://github.com/Toblerity/Fiona/issues/685) - this thread seems to be mostly about CPU but doesn't see to mention Memory consumption - is there a specific part there I've overlooked?

Also wondering if there is somewhere else in the docs or somewhere we've demonstrated empirically (e.g. via `psutil` as you suggest) that `.write` results in reduced memory consumption over `.writerecords`. If so it would be great to have these as part of the PR/repo, and documented in comments here for posterity (so we don't do something silly and revert the change as part of a refactor some years from now).

Here are a few of the links to some of the datasets that tend to be bottlenecks and intermittently crash depending on what else is causing load on CKAN (btw your suggestion to move this to data-pusher for this reason makes a ton of sense - thank you).
- https://open.toronto.ca/dataset/solarto/
- https://open.toronto.ca/dataset/police-annual-statistical-report-traffic-collisions/
- https://open.toronto.ca/dataset/street-tree-data/
- https://open.toronto.ca/dataset/topographic-mapping-building-outlines/

(Will take me a few more minutes to find a specific CSV from these that have caused the issue in the past - will get back to you here shortly!)

If we're having trouble reproducing we could also possibly try in a more resource-constrained environment - e.g. a docker container with a confined memory limitation (`docker run -m ...`)

Reply:
> Regarding that linked issue (https://github.com/Toblerity/Fiona/issues/685) - this thread seems to be mostly about CPU but doesn't seem to mention Memory consumption - is there a specific part there I've overlooked?

It seems to be the wrong link from the same author: https://github.com/Toblerity/Fiona/issues/624

> Also wondering if there is somewhere else in the docs or somewhere we've demonstrated empirically (e.g. via psutil as you suggest) that .write results in reduced memory consumption over .writerecords

I mentioned this as an option to test the difference between the two implementations since I couldn't replicate the system crashing.

I've tried the Solarto dataset from the list you shared, and I believe the others are smaller in size compared to it. However, I will test those as well.

> (Will take me a few more minutes to find a specific CSV from these that have caused the issue in the past - will get back to you here shortly!)

that will be good

> If we're having trouble reproducing we could also possibly try in a more resource-constrained environment - e.g. a docker container with a confined memory limitation (docker run -m ...)

I’ve actually done that. Currently, I’m using the following in my Docker Compose configuration:
```
deploy:
      resources:
        limits:
          memory: 6g
          cpus: "1.8"
```

Do you think that's too much for the work?

Reply:
thanks!

The thread mentions that a version bump of fiona to 1.8.0 resolved the memory consumption issue - we've been ahead of that version for a long time now though. Is there somewhere else `.write` and `.writerecords` memory tradeoffs are discussed? (not sure its necessary so long as we can prove it empirically just interested in the motivation)

> I mentioned this as an option to test the difference between the two implementations since I couldn't replicate the system crashing.

I think "testing the difference between the two implementations" is a good path forward if you agree! Seeing an empirical drop in memory usage from pre-change to post-change is probably more valuable than reproducing a full OOM situation (which depends on resources of the machine, other things running, etc..) anyway.

Perhaps we could add a dev script or unit test using `psutil` to perform some profiling to compare performance before and after these changes are applied to demonstrate that these changes reduce memory footprint? what do you think?

Re: CSVs - sorry for the delay here: 2 specific CSV that have caused issues in the past come from the solarto and police traffic collisions datasets, download links:
- https://ckan0.cf.opendata.inter.prod-toronto.ca/dataset/solarto/resource/6b49daae-8797-4153-b106-ccc404ec95b3/download/solarto-map%20-%204326.csv
- https://ckan0.cf.opendata.inter.prod-toronto.ca/dataset/ec53f7b2-769b-4914-91fe-a37ee27a90b3/resource/cb890861-ed20-4862-bb75-b1f9ec1e58dd/download/Traffic%20Collisions%20-%204326.csv

(but again I think I'm leaning more towards demonstrating a memory footprint improvement rather than reproducing a `MemoryError` if you agree - the docker resources you shared there look much less than our prod servers but its hard to say what else has been putting load on our servers at the times when we've run into memory errors).

Reply:
Well, the claim is not that `write` is better than `writerecords`; the claim is that there might be a memory improvement if one record is written at a time. Also, write is essentially just writerecords with one item in it.

I also did memory profiling. Writing a single record at a time uses less memory, but other factors consume more memory, which results in both implementations using much more memory, with only a small difference between them.

Here’s the note: https://hackmd.io/@steveoni/By03gSjGJx (you can jump to summary here https://hackmd.io/@steveoni/By03gSjGJx#Summary )

You can also try the script in your environment to profile the API.

Reply:
Thanks again Steve - these profile scripts look very helpful! will try them out in our environment and report back asap as we discussed

Reply:
@JamieBeverley here is an example of how you can add and use the script:

https://github.com/steveoni/ckanext-iotrans/commit/3443c8141279bc7d295a0be0f35d29b3126c7e75

you can use the same concept to wrap the code in your codebase.

and to run it, you can `sh` into your ckan pod and run this

```
mprof run --include-children --interval 0.1 ckan -c production.ini  ioprofile
```

also need to install `memory-profiler==0.61.0`